### PR TITLE
Bookstore tests

### PR DESCRIPTION
**Pull request by azzilian**
tests

Added Tests for BookController, CategoryController, BookService, and CategoryService classes

### CODE REVIEW

**Comment by azzilian** [comment]
Gonna add some more test later on

File: src/test/resources/database/book/remove-books-from-books-table.sql [hunk 1]
```diff
@@ -0,0 +1 @@
+DELETE FROM books WHERE Id BETWEEN 1 AND 2;
```
Comment (vikkushnir) [inline src/test/resources/database/book/remove-books-from-books-table.sql]:
```suggestion
DELETE FROM books WHERE Id BETWEEN 1 AND 2;

```

File: src/test/resources/database/book/remove-books_categories-from-books_categories-table.sql [hunk 2]
```diff
@@ -0,0 +1 @@
+DELETE FROM books_categories WHERE book_id BETWEEN 1 AND 2;
```
Comment (vikkushnir) [inline src/test/resources/database/book/remove-books_categories-from-books_categories-table.sql]:
```suggestion
DELETE FROM books_categories WHERE book_id BETWEEN 1 AND 2;

```

File: src/test/resources/database/book/remove-categories-from-categories-table.sql [hunk 3]
```diff
@@ -0,0 +1 @@
+DELETE FROM categories WHERE Id BETWEEN 1 AND 2;
```
Comment (vikkushnir) [inline src/test/resources/database/book/remove-categories-from-categories-table.sql]:
```suggestion
DELETE FROM categories WHERE Id BETWEEN 1 AND 2;

```

File: src/test/resources/database/book/remove-lenore-from-books-table.sql [hunk 4]
```diff
@@ -0,0 +1 @@
+DELETE FROM books WHERE title = 'Lenore' and price = 10.50;
```
Comment (vikkushnir) [inline src/test/resources/database/book/remove-lenore-from-books-table.sql]:
```suggestion
DELETE FROM books WHERE title = 'Lenore' and price = 10.50;

```

**Review by vikkushnir — APPROVED** [review]
Looks very nice to me! Only, add extra lines in your SQL files

File: src/test/java/com/onlinebookstore/onlinebookstore/controller/BookControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+                .setId(1L)
+                .setTitle(requestDto.getTitle())
+                .setAuthor(requestDto.getAuthor())
+                .setIsbn(requestDto.getIsbn())
+                .setPrice(requestDto.getPrice())
+                .setDescription(requestDto.getDescription())
+                .setCoverImage(requestDto.getCoverImage())
+                .setCategoryIds(requestDto.getCategoryIds());
+
+        Mockito.when(bookService.save(Mockito.any(BookRequestDto.class)))
```
Comment (andrii-hoienko) [inline src/test/java/com/onlinebookstore/onlinebookstore/controller/BookControllerTest.java]: Please use static imports

File: src/test/java/com/onlinebookstore/onlinebookstore/controller/BookControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        String jsonRequest = objectMapper.writeValueAsString(requestDto);
+
+        MvcResult result = mockMvc.perform(post("/api/books")
+                        .content(jsonRequest)
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isCreated())
+                .andReturn();
+
+        String jsonResponse = result.getResponse().getContentAsString();
+        System.out.println("Response JSON: " + jsonResponse);
```
Comment (andrii-hoienko) [inline src/test/java/com/onlinebookstore/onlinebookstore/controller/BookControllerTest.java]: Is it necessary to write something to console?

File: src/test/java/com/onlinebookstore/onlinebookstore/controller/BookControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+        assertTrue(EqualsBuilder.reflectionEquals(expected, actual, "id"));
+    }
+
+    @Test
+    @DisplayName("Get All Books - see all books in DB")
+    @WithMockUser(username = "admin", roles = {"ADMIN"})
+    void getAllBooks_ValidData_OK() throws Exception {
+        mockMvc.perform(get("/api/books"))
+                .andExpect(status().isOk());
```
Comment (andrii-hoienko) [inline src/test/java/com/onlinebookstore/onlinebookstore/controller/BookControllerTest.java]: Better to check content as well

File: src/test/java/com/onlinebookstore/onlinebookstore/controller/CategoryControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+    @SneakyThrows
+    private static void teardown(DataSource dataSource) {
+        try (Connection connection = dataSource.getConnection()) {
+            connection.setAutoCommit(true);
+            ScriptUtils.executeSqlScript(
+                    connection,
+                    new ClassPathResource("database/book/remove-all-from-books-table.sql")
+            );
+        }
+    }
```
Comment (andrii-hoienko) [inline src/test/java/com/onlinebookstore/onlinebookstore/controller/CategoryControllerTest.java]: Code duplication, it is better to create one teardown method that clear all db and call it when required

File: src/test/java/com/onlinebookstore/onlinebookstore/controller/CategoryControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        Mockito.when(categoryService.save(any(CategoryRequestDto.class)))
+                .thenReturn(categoryResponseDto);
+
+        mockMvc.perform(post("/api/categories")
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(objectMapper.writeValueAsString(categoryRequestDto)))
+                .andExpect(status().isCreated())
+                .andExpect(jsonPath("$.id", is(categoryResponseDto.id().intValue())))
+                .andExpect(jsonPath("$.name", is(categoryResponseDto.name())))
+                .andExpect(jsonPath("$.description", is(categoryResponseDto.description())));
```
Comment (andrii-hoienko) [inline src/test/java/com/onlinebookstore/onlinebookstore/controller/CategoryControllerTest.java]: Sometimes you check content using objectMapper, sometimes via json, please choose approach and use it everywhere to be consistent

File: src/test/java/com/onlinebookstore/onlinebookstore/service/impl/CategoryServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+        when(categoryRepository.findAll(pageable)).thenReturn(page);
+        when(categoryMapper.toDto(category)).thenReturn(responseDto);
+
+        List<CategoryResponseDto> result = categoryService.findAll(pageable);
+
+        verify(categoryRepository).findAll(pageable);
+        assertNotNull(result);
+        assertEquals(1, result.size());
+        assertEquals(category.getName(), result.get(0).name());
+        assertEquals(category.getDescription(), result.get(0).description());
```
Comment (andrii-hoienko) [inline src/test/java/com/onlinebookstore/onlinebookstore/service/impl/CategoryServiceImplTest.java]: check full object using only 1 assert

File: src/test/java/com/onlinebookstore/onlinebookstore/service/impl/BookServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+
+    @Test
+    @DisplayName("Delete book by id")
+    @Sql(scripts = {
+            "classpath:database/book/remove-all-from-books-table.sql"
+    }, executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
+    void deleteBookById_ValidData_OK() {
+        Book book = new Book();
+        book.setId(1L);
+        book.setTitle("Book 1");
```
Comment (andrii-hoienko) [inline src/test/java/com/onlinebookstore/onlinebookstore/service/impl/BookServiceImplTest.java]: Code duplication

Comment (andrii-hoienko) [reply src/test/java/com/onlinebookstore/onlinebookstore/service/impl/BookServiceImplTest.java]: Actually there are a lot of duplication when you create mock objects it makes sense to make a methods for that

File: src/test/java/com/onlinebookstore/onlinebookstore/service/impl/BookServiceImplTest.java [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+
+        assertNotNull(savedBookDto);
+        assertEquals(dto.getTitle(), savedBookDto.getTitle());
+    }
+
+    @Test
+    @DisplayName("Update existing Book")
+    @Sql(scripts = {
+            "classpath:database/book/remove-all-from-books-table.sql"
+    }, executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
```
Comment (andrii-hoienko) [inline src/test/java/com/onlinebookstore/onlinebookstore/service/impl/BookServiceImplTest.java]: Services tests should be unit, so we need to mock repository behavior. And test repo separatelly

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Nice, left some comments

File: src/test/java/com/onlinebookstore/onlinebookstore/controller/BookControllerTest.java (hunk 6 above)
Comment (azzilian) [reply src/test/java/com/onlinebookstore/onlinebookstore/controller/BookControllerTest.java]: oops, I have a lot of problems here so forgot to remove this one

File: src/test/java/com/onlinebookstore/onlinebookstore/controller/BookControllerTest.java [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+        assertTrue(EqualsBuilder.reflectionEquals(expected, actual, "id"));
+    }
+
+    @Test
+    @DisplayName("Get All Books - see all books in DB")
+    @WithMockUser(username = "admin", roles = {"ADMIN"})
+    void getAllBooks_ValidData_OK() throws Exception {
+        mockMvc.perform(get("/api/books")
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk());
```
Comment (andrii-hoienko) [inline src/test/java/com/onlinebookstore/onlinebookstore/controller/BookControllerTest.java]: image

File: src/test/java/com/onlinebookstore/onlinebookstore/controller/BookControllerTest.java [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    @Test
+    @DisplayName("Find Book by Id")
+    @WithMockUser(username = "admin", roles = {"ADMIN"})
+    void findBookById_ValidData_OK() throws Exception {
+        mockMvc.perform(get("/api/books/{id}", BOOK_ID))
+                .andExpect(status().isOk());
```
Comment (andrii-hoienko) [inline src/test/java/com/onlinebookstore/onlinebookstore/controller/BookControllerTest.java]: image

File: src/test/java/com/onlinebookstore/onlinebookstore/controller/BookControllerTest.java [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+
+    @Test
+    @DisplayName("Update Book")
+    @WithMockUser(username = "admin", roles = {"ADMIN"})
+    void updateBookById_ValidData_OK() throws Exception {
+        String jsonRequest = objectMapper.writeValueAsString(requestDto);
+        mockMvc.perform(put("/api/books/{id}", BOOK_ID)
+                        .content(jsonRequest)
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk());
```
Comment (andrii-hoienko) [inline src/test/java/com/onlinebookstore/onlinebookstore/controller/BookControllerTest.java]: image

File: src/test/java/com/onlinebookstore/onlinebookstore/service/impl/BookServiceImplTest.java [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+
+        pageable = PageRequest.of(0, 10);
+    }
+
+    @Test
+    @DisplayName("Find all books in database")
+    void findAllBooks_ValidData_OK() {
+        Page<Book> bookPage = new PageImpl<>(Arrays.asList(book1, book2));
+
+        Mockito.when(bookRepository.findAll(pageable)).thenReturn(bookPage);
```
Comment (andrii-hoienko) [inline src/test/java/com/onlinebookstore/onlinebookstore/service/impl/BookServiceImplTest.java]: image

File: src/test/java/com/onlinebookstore/onlinebookstore/service/impl/BookServiceImplTest.java [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+        assertThat(actualMessage).contains(EXPECTED_MESSAGE);
+    }
+
+    @Test
+    @DisplayName("Create new Book")
+    void createNewBook_ValidData_OK() {
+        Mockito.when(categoryRepository.findAllById(categoryIds)).thenReturn(categories);
+        Mockito.when(bookMapper.toModel(bookRequestDto)).thenReturn(book1);
+        Mockito.when(bookRepository.save(book1)).thenReturn(book1);
+        Mockito.when(bookMapper.toDto(book1)).thenReturn(bookResponseDto);
```
Comment (andrii-hoienko) [inline src/test/java/com/onlinebookstore/onlinebookstore/service/impl/BookServiceImplTest.java]: image

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Please be cerafull, and check comment once again

File: src/test/java/com/onlinebookstore/onlinebookstore/controller/CategoryControllerTest.java [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+    @Test
+    @DisplayName("Get category by ID")
+    @WithMockUser(username = "admin", roles = {"ADMIN"})
+    void getCategoryById_OK() throws Exception {
+        when(categoryService.getById(anyLong())).thenReturn(categoryResponseDto);
+
+        mockMvc.perform(get("/api/categories/{id}",
+                        CATEGORY_ID_FIRST)
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk());
```
Comment (karma-o) [inline src/test/java/com/onlinebookstore/onlinebookstore/controller/CategoryControllerTest.java]: please add response body assertions for this class as well

**Review by karma-o — APPROVED** [review]
Left a minor comment, other than that, it looks good :)

**Pull request merged**